### ISSUE TITLE
Built-in types should be keywords

### ISSUE THREAD

**Issue by RalfJung** [integer32llc/rust-playground#172]
Please consider marking the built-in types (`i32` and friends) as keywords. IIRC they were keywords in the old playground, and they are keywords in most (all?) editors with Rust syntax highlighting.

**Comment by shepmaster** [comment]
> the built-in types

Any idea what the deciding line is for builtins? Things in libstd, libcore? Is there a nicely prepared list for me somewhere? :-)

**Comment by RalfJung** [comment]
I was referring to the "lower-case" types -- `i*`, `u*`, `f*`.

**Comment by shepmaster** [comment]
> the built-in types (i32 and friends) as keywords

My first thoughts are that I disagree with this. If it's a *type*, it's **not** a *keyword*. If anything, the lowercase types should be highlighted the same as "regular" types.

> and they are keywords in most (all?) editors

Here's my emacs setup, FWIW:

![image](https://user-images.githubusercontent.com/174509/28219259-5c902662-6889-11e7-8a49-45feec439316.png)

Note that the type colors match up and are distinct from the keyword colors.

**Comment by RalfJung** [comment]
> If it's a type, it's not a keyword. I

I don't agree with this in general, there are certainly languages that do make their primitive types keywords. Rust, however -- to my surprise -- permits `let i32 = 0;`. oO

> Note that the type colors match up and are distinct from the keyword colors.

That's not what I see -- I see enum variants also in the same color, which are not types. (That is assuming `Assembly` is an enum variant; it could also be a binding variable name.) In general, detecting whether something is a type needs a full semantic analysis (e.g., syntactically speaking, `CompileTarget` above could be a module or a type), so I felt asking for that would be asking too much :)

In the end, I don't care whether they look like keywords or something else, but I think it'd be helfpul for them to be different in *some* way from "mere" identifiers.

**Comment by shepmaster** [comment]
The editor actually does parse these differently, it's up to the theme to decide how to show those. For example, here is `mono_industrial`:

![image](https://user-images.githubusercontent.com/174509/28223644-4072d734-689a-11e7-99f9-d9d62f731e11.png)

Perhaps there's a theme that better fits your intent?

**Comment by RalfJung** [comment]
I did not even know that there are different themes I could switch between... I guess there will be at least one in that list that I like. Of course, I don't permit websites to store anything across a browser restart, so this won't last long. Oh well.

So the default theme is called "github" -- it looks fairly different from github though.
Notice that github *does* highlight these types:
```rust
fn main() {
  let x: i32 = 2;
}
```

**Comment by shepmaster** [comment]
> So the default theme is called "github" -- it looks fairly different from github though.

Yeah, that's being provided by [Ace](https://ace.c9.io/). It looks like their themes [aren't really updated all that frequently, though](https://github.com/ajaxorg/ace/tree/master/lib/ace/theme). 

Two possibilities I could see:

1. Someone (likely not me) submits an update to Ace to update the `github` style.
1. Someone (likely not me) creates a brand new Ace theme that we use as the default playground style.

---

Note that GitHub *does* highlight both `i32` and `String`:

```rust
fn main() {
  let x: i32 = 2;
  let x: String = 2;
}
```

But doesn't try "as hard" as my emacs setup:

```rust
use self::CompileTarget::*;

match *self {
    Assembly => "assembly".fmt(f),
```

**Comment by shepmaster** [comment]
> I don't permit websites to store anything across a browser restart, so this won't last long.

I can't help much with that, but it's possible you could have some sort of browserscript which you control that always applies the theme you want (or whitelist the domain; you do trust me, don't you? 😈)
